[zhuuyicheng] iP

zhuuyicheng opened the pull request:
# Zack

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

The Zack chatbot frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://github.com/zhuuyicheng/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the ``main`` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

zhuuyicheng committed:
Level-0

zhuuyicheng committed:
Level-1. Echo

zhuuyicheng committed:
Level-2. Add, List

zhuuyicheng committed:
Added Task.java to implement A-Classes

zhuuyicheng committed:
Level 4. ToDos, Events, Deadlines. Abstracted adding tasks into a method.

zhuuyicheng committed:
A-TextUiTesting: Automated Text UI Testing

zhuuyicheng committed:
Level-5. Handle Errors.
Implemented a custom Exception class, abstracted away many of the if/else statements into methods.
Included multiple handle errors with unique messages for each specific error.

zhuuyicheng committed:
Updated Automated Text UI Testing cases

zhuuyicheng committed:
Level 6. Delete
Included error handling for Delete
Updated Automated Text UI Testing

zhuuyicheng committed:
A-Enums
Incoporated the use of Java enums to help with switch/case statements

zhuuyicheng committed:
Unindent case statements

zhuuyicheng committed:
Implement save function of chatbot in the Storage class

zhuuyicheng committed:
Merge branch 'branch-Level-7'

zhuuyicheng committed:
Store deadline and event dates and times in standardised formats

zhuuyicheng committed:
Add DATE command to print deadlines/events occurring on a specific date

zhuuyicheng committed:
Merge branch 'branch-Level-8'

zhuuyicheng committed:
Extract methods in Zack.java into Ui, Storage, Parser and TaskList classes to make the code more OOP

zhuuyicheng committed:
Merge tag 'A-MoreOOP'

zhuuyicheng committed:
Divide classes into multiple packages

zhuuyicheng committed:
Merge branch 'A-Packages'

zhuuyicheng committed:
Merge remote-tracking branch 'origin/add-gradle-support'

zhuuyicheng committed:
Set up gradle so it can build and run Zack

zhuuyicheng committed:
Merge branch 'A-Gradle'

zhuuyicheng committed:
Add JUnit to parse method in Parser class as well as for parseTask in Storage class, both of which are tested reasonably well

zhuuyicheng committed:
Merge branch 'A-JUnit'

zhuuyicheng committed:
Set up gradle to package the app as an executable JAR file

zhuuyicheng committed:
Merge branch 'A-Jar'

zhuuyicheng committed:
Add JavaDoc comments to all non-private classes/methods, and non-trivial private methods

zhuuyicheng committed:
Tweak the code to comply with the given coding standard

zhuuyicheng committed:
Implement the find function to allow users to find a task by searching for a keyword

zhuuyicheng committed:
Merge branch 'branch-A-JavaDoc'

zhuuyicheng committed:
Merge branch 'A-CodingStandard'

# Conflicts:
#	src/main/java/zack/commands/DeleteCommand.java

zhuuyicheng committed:
Merge branch 'branch-Level-9'

zhuuyicheng committed:
Use checkStyle to detect coding style violations

zhuuyicheng committed:
Merge branch 'A-CheckStyle'

zhuuyicheng committed:
Implement iteration 2 by adding dialog boxes

zhuuyicheng committed:
Implement Iteration 3 - Adding custom behavior

zhuuyicheng committed:
Rebuild the scene using FXML

zhuuyicheng committed:
Merge branch 'Level-10'

zhuuyicheng committed:
Merge branch 'master' into Level-10

zhuuyicheng committed:
Implement Level-10 by adding a GUI to the chatbot

zhuuyicheng committed:
Merge branch 'Level-10'

leejaegeun-sql commented on src/main/java/zack/util/Ui.java:
```diff
@@ -0,0 +1,199 @@
+    /**
+     * Prints a line divider with an additional new line.
+     */
+    public void showNewLineLine() {
```
Is it better to name this showNewLine() instead?

leejaegeun-sql commented on src/test/java/zack/ParserTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Test
+    public void parse_deadlineCommandWithDescriptionAndDate_addTaskCommandReturned() throws Exception {
```
Would it better if the name of the the function could be phrased shorter?

leejaegeun-sql commented on src/test/java/zack/ParserTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Test
+    public void parse_todoCommandWithDescription_addTaskCommandReturned() throws Exception {
```
Would it better if the name of the the function could be phrased shorter?

leejaegeun-sql commented on src/test/java/zack/ParserTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Test
+    public void parse_eventCommandWithDescriptionAndDate_addTaskCommandReturned() throws Exception {
```
Would it better if the name of the the function could be phrased shorter?

leejaegeun-sql commented on src/test/java/zack/ParserTest.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @Test
+    public void parse_deleteCommandWithNumericIndex_deleteCommandReturned() throws Exception {
```
Would it better if the name of the the function could be phrased shorter?

leejaegeun-sql reviewed (commented):
LGTM - very minor points to consider in order to improve the code quality

zhuuyicheng committed:
Remove prefHeight so height of dialog box can be adjusted dynamically

zhuuyicheng committed:
Use assert feature to document assumptions

Assertions serve as a form of self-documentation, clarifying the expected state or behavior at
various points in the program. This enhances code readability and maintainability.

The assertions have been mostly added in classes inside the Util folder as they
rely on the most assumptions.

zhuuyicheng committed:
Improve code quality in accordance to the textbook

Redundant UI code, long method names, and overly complex methods
can affect a codebase's maintainability and readability.

Lets remove redundant UI code to streamline the application.
This helps to eliminate unnecessary complexity and focuses
on the essential features, making the codebase easier to maintain
and extend.

Let's rename method names that are excessively long to
be concise yet descriptive. This change improves readability
and simplifies navigation within the code.

Let's reduce method length to allow for ease of
locating and addressing potential issues in the
future.

zhuuyicheng committed:
Merge pull request #1 from zhuuyicheng/A-Assertions

Use assert feature to document assumptions

zhuuyicheng committed:
Revert "Use assert feature to document assumptions"

zhuuyicheng committed:
Merge pull request #2 from zhuuyicheng/revert-1-A-Assertions

Revert "Use assert feature to document assumptions"

zhuuyicheng committed:
Revert "Revert "Use assert feature to document assumptions""

zhuuyicheng committed:
Merge pull request #4 from zhuuyicheng/revert-2-revert-1-A-Assertions

Revert "Revert "Use assert feature to document assumptions""

zhuuyicheng committed:
Merge branch 'master' of https://github.com/zhuuyicheng/ip

zhuuyicheng committed:
Merge branch 'master' into A-CodeQuality

STELLA-LYE commented on src/test/java/zack/StorageTest.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    @Test
+    void parseTask_deadlineFormat_parsedSuccessfully() throws ZackException {
+        String input = "D | 1 | submit assignment | 2022-12-25 2200";
+        Task task = storage.parseTask(input);
+        assertTrue(task instanceof Deadline);
+        assertTrue(task.isDone());
+        assertEquals("submit assignment", task.getDescription());
+    }
+
+    @Test
+    void parseTask_eventFormat_parsedSuccessfully() throws ZackException {
```
Would it be better to give the functions short and concise name?

STELLA-LYE reviewed (commented):
LGTM.  I found your code easy to read and well-written. Just a few nits to consider.

STELLA-LYE commented on src/main/java/zack/Zack.java:
```diff
@@ -0,0 +1,239 @@
+    private TextField userInput;
+    private Button sendButton;
+    private Scene scene;
+    private Image user = new Image(this.getClass().getResourceAsStream("/images/user.png"));
```
Perhaps can use a more detailed naming like userImage?

STELLA-LYE commented on src/main/java/zack/Zack.java:
```diff
@@ -0,0 +1,239 @@
+    private Button sendButton;
+    private Scene scene;
+    private Image user = new Image(this.getClass().getResourceAsStream("/images/user.png"));
+    private Image zack = new Image(this.getClass().getResourceAsStream("/images/zack.jpg"));
```
Perhaps can use a more detailed naming like zackImage?

STELLA-LYE commented on src/main/java/zack/tasks/Task.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return True if the task is marked as done, false otherwise.
+     */
+    public boolean isDone() {
```
I like how there is detailed function names and a clear boolean function name.

STELLA-LYE commented on src/main/java/zack/util/Ui.java:
```diff
@@ -0,0 +1,205 @@
+     * @throws ZackException If there is an error or exception while displaying
+     *                      the found tasks.
+     */
+    public String showFoundTasks(TaskList foundTasks) throws ZackException {
```
I like how the function names are detailed and self-explanatory. Makes the code easy to understand.

STELLA-LYE reviewed (commented):
LGTM.  I found your code easy to read and well-written.

zhuuyicheng committed:
Merge pull request #3 from zhuuyicheng/A-CodeQuality

Improve code quality in accordance to the textbook

zhuuyicheng committed:
Merge branch 'master' of https://github.com/zhuuyicheng/ip

zhuuyicheng committed:
Remove redundant code and enable java assertions

zhuuyicheng committed:
Merge branch 'minor-fixes'

zhuuyicheng committed:
Remove the github workflows file

zhuuyicheng committed:
Use GitHub Actions to set up Continuous Integration (CI)

zhuuyicheng committed:
Merge branch 'master' of https://github.com/zhuuyicheng/ip

zhuuyicheng committed:
Add extension to sort tasks

Implement streams in the extension

zhuuyicheng committed:
Merge pull request #5 from zhuuyicheng/C-Sort

Add extension to sort tasks

zhuuyicheng committed:
Merge branch 'C-Sort'

zhuuyicheng committed:
Merge branch 'master' of https://github.com/zhuuyicheng/ip

zhuuyicheng committed:
Enlist the use of AI to write documentation for methods

zhuuyicheng committed:
Merge branch 'A-AiAssisted'

zhuuyicheng committed:
Give a unique personality in the way Zack responses

zhuuyicheng committed:
Merge branch 'A-Personality'

zhuuyicheng committed:
More error handling

Handle unexpected input such as when start date/time is later(or same as) than end/time, and a value that should be unique is duplicated (e.g., two tasks with the same exact details).

zhuuyicheng committed:
Merge branch 'A-MoreErrorHandling'

zhuuyicheng committed:
Add a user guide for Zack chatbot

zhuuyicheng committed:
Merge branch 'A-UserGuide'

zhuuyicheng committed:
Add Zack chatbot screenshot in user guide

zhuuyicheng committed:
Update initial README on github

zhuuyicheng committed:
Improve the GUI to make it more polished

zhuuyicheng committed:
Merge branch 'A-BetterGui'

zhuuyicheng committed:
Add link to github release in README

zhuuyicheng committed:
Update formatting of user guide

zhuuyicheng committed:
Improve clarity of README

zhuuyicheng committed:
Fix error message wording

zhuuyicheng committed:
Add title to application title bar

zhuuyicheng committed:
Allow app to close when user inputs "bye"

zhuuyicheng committed:
Make the application window unresizable
